Break ScheduledMessage out into OnboardingMessage and QuarterlyMessage

jmhooper opened the pull request:
Fixes issue(s) #206

Changes proposed in this pull request:

- Break `ScheduledMessages` model into `QuarterlyMessage` model and `OnboardingMessage` model
- Change `SentScheduledMessage` model to `SentMessage` and make the relationship to messages polymorphic

This PR is WIP

jessieay commented on app/assets/javascripts/scheduled_message.js:
```diff
@@ -1,15 +0,0 @@
-$(document).ready(function() {
```
yess so happy to be done with this.

jmhooper replied:
👋

jessieay commented on app/models/quarterly_message.rb:
```diff
@@ -0,0 +1,30 @@
+  validates :title, presence: true
+
+  # disable STI
+  self.inheritance_column = nil
```
do we need this now that there is no longer a `type` column?

jmhooper replied:
We do not. I'll clean up.

jessieay commented on db/chores/migrate_messages_to_scheduled_messages.rb:
```diff
@@ -0,0 +1,91 @@
+  def migrate_onboarding_message_to_scheduled_message(row)
+    # Migrate to new table
+    row["type"] = 0
+    values = ["created_at","updated_at","title","body","days_after_start","time_of_day","type","end_date","deleted_at"].map do |column|
```
small style thing: would it be possible to put each array val on its own line? I think that would be easier to read

jmhooper replied:
I can. The reason I did this was so that in dev I could line them up visually with the column names in the SQL a couple lines down.

jessieay replied:
fine as-is if you prefer it!

jessieay commented on db/chores/migrate_messages_to_scheduled_messages.rb:
```diff
@@ -0,0 +1,91 @@
+      SQL
+    )
+
+    # Update sent messages
```
what about putting each of these into a well-named method rather than using comments? eg: lines 27 through 33 can be `def update_sent_messages` ?

jessieay commented on db/chores/migrate_messages_to_scheduled_messages.rb:
```diff
@@ -0,0 +1,91 @@
+
+  # Helpers
+
+  def execute(*args)
```
not necessary but it might be good to put these into a shared place since you use them in two places 🌞

jmhooper replied:
I thought about that. Maybe something like a `ScheduledMessageSQLMethods` module that both of these could include?

jessieay replied:
sure, sounds great

jessieay commented on spec/models/onboarding_message_spec.rb:
```diff
@@ -0,0 +1,35 @@
+
+  describe '.date_time_ordering' do
+    it 'should display the messages in chronological order' do
+      m1 = create(:onboarding_message, days_after_start: 1, time_of_day: '2000-01-01 11:00:00 UTC')
```
can we create these out of order so we know that the ordering is working?  I think the would return in this order by default anyway so that would increase the quality of test

jmhooper replied:
Yes, good idea

jmhooper commented:
> I'd love to see some unit tests for the chores classes so we have confidence that the migrate data the way we want them to!

This was tripping me up the other night. The chores require the schema to be in a weird transitional state to work. For example, to migrate the scheduled messages to quarterly/onboarding messages, you need the `scheduled_messages` table, the `quarterly_messages` table, and the `onboarding_messages` table. The schema is not in that state before or after the migrations.

The migration tests I was looking at involved rolling back to a particular migration, adding some data, running the next migration, and testing the result. I'm not sure how to do something like that for the chores since necessary state for the schema isn't achievable by rolling back to a particular migration.

My thought was to roll back to the previous migration, grab the code from the `up` and `down` methods of the migration before the chores are called, and move it into a `before_each` hook in spec. I'm not a huge fan of that solution though b/c it seems like there should be a Better Way™

Any thoughts?

jessieay commented:
@jmhooper ah yes, that is super complicated. Did you try out the technique described in http://blog.carbonfive.com/2011/01/27/start-testing-your-migrations-right-now/ (which you linked me to)? 

I've never used something like that before, but looks interesting! It also looks like the strategies they describe involve running the migration. I think that if you went back to the previous migration version (`ActiveRecord::Migrator.migrate @previous_version` in that blog post), created some records with the old DB schema, then ran the migration and confirmed that the new records that you'd expect exist, that would be more than good enough w/r/t testing.

jmhooper commented:
@jessieay: Addressed some of your comments and pushed some more commits to round this out.

The status of this can change from WIP to Ready for Review.

jessieay commented on app/models/onboarding_message.rb:
```diff
@@ -0,0 +1,38 @@
+    order(:days_after_start, :time_of_day)
+  end
+
+  def self.filter(params)
```
since this is exactly the same here and in `QuarterlyMessage` -- thoughts on putting it in a shared concern? http://api.rubyonrails.org/classes/ActiveSupport/Concern.html

jmhooper replied:
My thought was to open an issue once this was merged to move it into a query object. I can go ahead and do that or move it into a concern in this PR though.

jessieay replied:
sure! good plan, open an issue.

jessieay commented on app/models/sent_message.rb:
```diff
@@ -0,0 +1,54 @@
+    where("extract(year from created_at) = ?", year)
+  end
+
+  def self.filter(params)
```
some comment here re: moving shared logic into a concern

jessieay commented on app/views/onboarding_messages/edit.html.erb:
```diff
@@ -0,0 +1,6 @@
+<%= render 'header', :title => 'Update onboarding message' %>
+
+<section class="main-content">
+	<%= render "flashes" -%>
```
maybe flashes should be in layout? (this could be done via a separate issue/PR)

jmhooper replied:
I agree. I can open an issue.
